torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -870,6 +872,120 @@ def _check_supported_callable(fn_var):
         )
 
 
+class AssociativeScanHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    @raise_hard_error_if_graph_break(
+        reason="associative_scan must be captured completely with torch.compile."
+    )
+    def call_function(
+        self, tx, args: List[VariableTracker], kwargs: Dict[str, VariableTracker]
+    ) -> VariableTracker:
+        from . import NestedUserFunctionVariable, UserFunctionVariable
+        from .builder import wrap_fx_proxy
+
+        args, kwargs = LazyVariableTracker.realize_all((args, kwargs))
+
+        def arg_extractor(input, dim, combine_fn):
+            return input, dim, combine_fn
+
+        input, dim, combine_fn = arg_extractor(*args, **kwargs)
+
+        assert isinstance(
+            combine_fn,
+            (
+                UserFunctionVariable,
+                NestedUserFunctionVariable,
+            ),
+        )
+
+        # operands
+        if input.python_type() != torch.Tensor:
+            unimplemented(
+                f"Expected input to be a tensor but got {input.python_type()}",
+            )
+
+        input_meta = input.as_proxy().node.meta["example_value"]
+        with tx.fake_mode:
+            example_vals = [
+                torch.empty((), device=input_meta.device, dtype=input_meta.dtype)
+                for _ in range(2)
+            ]
+
+        # Trace the subgraph
+        # NOTE: We don't use speculate_subgraph as it expects subgraph arguments to be
+        # variables in the current graph.
+        with tx.output.subtracer(self.value, prior_tracer=None) as subtracer:
```

Comment:
I would recommend using speculate_subgraph (or figuring out how to get speculate_subgraph to work). speculate_subgraph has some additional things (and we'll continue to build on top of it):
- side effect restoration: https://github.com/pytorch/pytorch/blob/8c7d8f0ff20b71a2dc1fbe976ee457267ab385a9/torch/_dynamo/variables/higher_order_ops.py#L394
- pytree handling

Reply:
This is how we're doing it for score_mod: https://github.com/pytorch/pytorch/pull/121845/files#diff-d6c3a57794a9a8a79eacedc57d68ff631fe80776f8701398eb3ef349973b033bR1306

Reply:
I was actually unaware it would produce extra junk 🤔 Is that right @drisspg ?

Reply:
I have only been looking at at the joint graph in aot which seemed to be correct to me - with no extra nodes, this may be the case though for what dynamo produces

Reply:
It produces extra junk, but:
1) it gets cleaned up later on in the PT2 stack
2) we can also just go back in and remove it manually.

In general we want to centralize on calling speculate subgraph (or something else!) so that future infra improvements to the subgraph capture mechanism will benefit all users.

Reply:
Okay, I've changed it to use `speculate_subgraph`.